src/v/pandaproxy/schema_registry/util.h

```diff
@@ -60,4 +63,44 @@ make_schema_definition(std::string_view sv) {
       ss::sstring{str_buf.GetString(), str_buf.GetSize()}};
 }
 
+template<
+  typename Service,
+  typename Func,
+  typename... Args,
+  typename Ret = ss::futurize_t<std::invoke_result_t<Func, Service&, Args...>>>
+requires std::invocable<Func, Service&, Args&&...>
+static Ret invoke_cross_shard_service_call(
+  ss::sharded<Service>& service,
+  ss::shard_id shard_id,
+  ss::smp_submit_to_options options,
+  Func&& func,
+  Args&&... args) {
+    // If running single core, the service group semaphore has a count of 0
+    if (ss::smp::count > 1) {
+        const auto& sg_sem = ss::get_smp_service_groups_semaphore(
+          ss::internal::smp_service_group_id(options.service_group), shard_id);
+        if (sg_sem.available_units() <= 0) {
+            throw as_exception(exhausted_cross_shard_semaphore());
+        }
+    }
+
+    return service.invoke_on(
+      shard_id, options, std::forward<Func>(func), std::forward<Args>(args)...);
+}
+
+template<typename Func>
+static ss::futurize_t<std::invoke_result_t<Func>> submit_cross_shard_call(
+  ss::shard_id shard_id, ss::smp_submit_to_options options, Func&& func) {
+    // If running single core, the service group semaphore has a count of 0
+    if (ss::smp::count > 1) {
+        const auto& sg_sem = ss::get_smp_service_groups_semaphore(
+          ss::internal::smp_service_group_id(options.service_group), shard_id);
+        if (sg_sem.available_units() <= 0) {
+            throw as_exception(exhausted_cross_shard_semaphore());
```

Comment:
Would it be possible to set `smp_submit_to_options::timeout` and handle `semaphore_timed_out` instead of wrapping everything?